index.js

```diff
@@ -1,207 +1,185 @@
-          })
-        }
-      })
-    }, (err, files) => {
-      if (err) {
-        next(err)
-        return
+    if (dirent.isFile() && scriptPattern.test(dirent.name)) {
+      const type = getScriptType(file, options.packageType)
+      if (type === 'typescript' && !typescriptSupport) {
+        throw new Error(`fastify-autoload cannot import plugin at '${file}'. To fix this error compile TypeScript to JavaScript or use 'ts-node' to run your app.`)
       }
-
-      const stats = [].concat(...files)
-
-      const allPlugins = {}
-
-      for (let i = 0; i < stats.length; i++) {
-        const { skip, file, opts } = stats[i]
-
-        if (skip) {
-          continue
-        }
-
-        try {
-          const content = require(file)
-          let plugin
-          if (content &&
-            Object.prototype.toString.apply(content) === '[object Object]' &&
-            Object.prototype.hasOwnProperty.call(content, 'method')) {
-            plugin = function (fastify, opts, next) {
-              fastify.route(content)
-              next()
-            }
-          } else {
-            plugin = content
-          }
-          const pluginConfig = (plugin.default && plugin.default.autoConfig) || plugin.autoConfig || {}
-          const pluginOptions = Object.assign({}, pluginConfig, defaultPluginOptions)
-          const pluginMeta = plugin[Symbol.for('plugin-meta')] || {}
-          const pluginName = pluginMeta.name || file
-
-          if (plugin.default && plugin.default.autoConfig && typeof plugin.default.autoConfig === 'object') {
-            plugin.default.autoConfig = undefined
-          }
-
-          if (typeof plugin.autoConfig === 'object') {
-            plugin.autoConfig = undefined
-          }
-
-          if (opts && !plugin.autoPrefix) {
-            plugin.autoPrefix = opts.prefix
-          }
-
-          if (plugin.autoload === false) {
-            continue
-          }
-
-          if (plugin.autoPrefix) {
-            const prefix = pluginOptions.prefix || ''
-            pluginOptions.prefix = prefix + plugin.autoPrefix
-          }
-
-          if (plugin.prefixOverride !== undefined) {
-            pluginOptions.prefix = plugin.prefixOverride
-          }
-
-          if (allPlugins[pluginName]) {
-            throw new Error(`Duplicate plugin: ${pluginName}`)
-          }
-
-          allPlugins[pluginName] = {
-            plugin,
-            name: pluginName,
-            dependencies: pluginMeta.dependencies,
-            options: pluginOptions
-          }
-        } catch (err) {
-          next(enrichError(err))
-          return
-        }
+      if (type === 'module' && !moduleSupport) {
+        throw new Error(`fastify-autoload cannot import plugin at '${file}'. Your version of node does not support ES modules. To fix this error upgrade to Node 14 or use CommonJS syntax.`)
       }
+      accumulator.push({ file, type, prefix })
+    }
+  }
+  await Promise.all(directoryPromises)
```

Comment:
if we use `Promise.anySettled` here we can collect all failures instead of failing on the first, might be useful

Reply:
We can propagate only one error up anyway. What should we do with those errors?